File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {

        Arrays.sort(nums);

        int length=nums.length;

        List<List<Integer>> ans=new ArrayList<>();

        for(int i=0;i<length-2;i++){

            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }

            int low=i+1;
            int high=length-1;
            while(low<high){
                if(nums[i]+nums[low]+nums[high]==0){
                    List<Integer> list=new ArrayList<>();
                    list.add(nums[i]);
                    list.add(nums[low]);
                    list.add(nums[high]);
                    ans.add(list);
                    low=low+1;
                    high=high-1;
                    while(low<high && nums[low]==nums[low-1]){
                        low=low+1;
                    }
                    while(low<high && nums[high]==nums[high+1]){
                        high=high-1;
                    }
                }else if((nums[i]+nums[low]+nums[high])<0){
                    low=low+1;
                }else{
                    high=high-1;
                }
            }

        }


        return ans;
        
    }
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans=[]
        nums.sort()
        for i in range(len(nums)-2):
            if i>0 and nums[i]==nums[i-1]:
                continue
            lo=i+1
            hi=len(nums)-1
            while lo<hi:
                if nums[i]+nums[lo]+nums[hi]==0:
                    ans.append([nums[i],nums[lo],nums[hi]])
                    lo=lo+1
                    hi=hi-1
                    while lo<hi and nums[lo]==nums[lo-1]:lo=lo+1
                    while lo<hi and nums[hi]==nums[hi+1]:hi=hi-1
                elif nums[i]+nums[lo]+nums[hi]>0:
                    hi=hi-1
                else:
                    lo=lo+1
        return ans
        